src/ssort/_files.py

```diff
@@ -1,72 +1,33 @@
-from __future__ import annotations
+from pathlib import Path
 
-import os
-import pathlib
-from typing import Iterable
+__all__ = ["find_project_root"]
 
-import pathspec
 
-from ssort._utils import memoize
+def current_working_dir():
+    return Path(".").resolve()
 
-_EMPTY_PATH_SPEC = pathspec.PathSpec([])
 
+def find_project_root(patterns):
+    all_patterns = [current_working_dir()]
 
-@memoize
-def _is_project_root(path: pathlib.Path) -> bool:
-    if path == path.root or path == path.parent:
-        return True
+    if patterns:
+        all_patterns.extend(patterns)
 
-    if (path / ".git").is_dir():
-        return True
+    paths = [Path(p).resolve() for p in all_patterns]
+    parents_and_self = [
+        list(reversed(p.parents)) + ([p] if p.is_dir() else []) for p in paths
+    ]
 
-    return False
+    *_, (common_base, *_) = (
+        common_parent
+        for same_lvl_parent in zip(*parents_and_self)
+        if len(common_parent := set(same_lvl_parent)) == 1
+    )
 
+    for directory in (common_base, *common_base.parents):
+        if (directory / ".git").exists() or (
+            directory / "pyproject.toml"
+        ).is_file():
+            return directory
 
-@memoize
-def _get_ignore_patterns(path: pathlib.Path) -> pathspec.PathSpec:
-    git_ignore = path / ".gitignore"
-    if git_ignore.is_file():
-        with git_ignore.open() as f:
-            return pathspec.PathSpec.from_lines("gitwildmatch", f)
-
-    return _EMPTY_PATH_SPEC
-
-
-def is_ignored(path: str | os.PathLike) -> bool:
-    # Can't use pathlib.Path.resolve() here because we want to maintain
-    # symbolic links.
-    path = pathlib.Path(os.path.abspath(path))
-
-    for part in (path, *path.parents):
-        patterns = _get_ignore_patterns(part)
-        if patterns.match_file(path.relative_to(part)):
-            return True
-
-        if _is_project_root(part):
-            return False
-
-    return False
-
-
-def find_python_files(
-    patterns: Iterable[str | os.PathLike[str]],
-) -> Iterable[pathlib.Path]:
-    if not patterns:
-        patterns = ["."]
-
-    paths_set = set()
-    for pattern in patterns:
-        path = pathlib.Path(pattern)
-        if not path.is_dir():
-            subpaths = [path]
-        else:
-            subpaths = [
-                subpath
-                for subpath in path.glob("**/*.py")
-                if not is_ignored(subpath) and subpath.is_file()
-            ]
-
-        for subpath in sorted(subpaths):
-            if subpath not in paths_set:
-                paths_set.add(subpath)
-                yield subpath
+    return common_base
```

Comment:
I think it's important for the git root to be per-path.  If it isn't then the results of running `ssort ./repo1/ ./repo2/` won't match the results of running `ssort ./repo1; ssort ./repo2/`.  Having a single git root will also break with git submodules.